mllam_data_prep/ops/cropping.py

```diff
@@ -0,0 +1,201 @@
+        chull_lam.contains_lonlat, da_lon.load(), da_lat.load(), vectorize=True
+    ).astype(bool)
+    da_interior_mask.attrs[
+        "long_name"
+    ] = "contained in convex hull of source dataset (da_ref)"
+
+    return da_interior_mask
+
+
+def _latlon_to_unit_sphere_xyz(
+    da_lat: xr.DataArray, da_lon: xr.DataArray
+) -> xr.DataArray:
+    """
+    Convert lat/lon coordinates to (x, y, z) on the unit sphere.
+
+    Parameters
+    ----------
+    da_lat : xarray.DataArray
+        Latitude coordinates.
+    da_lon : xarray.DataArray
+        Longitude coordinates.
+
+    Returns
+    -------
+    xr.DataArray
+        The (x, y, z) coordinates on the unit sphere as an xarray.DataArray
+        with dimensions (grid_index, component).
+    """
+    pts_xyz = np.array(sg.vector.lonlat_to_vector(da_lon, da_lat)).T
+    da_xyz = xr.DataArray(
+        pts_xyz, coords=da_lat.coords, dims=list(da_lat.dims) + ["xyz"]
+    )
+    return da_xyz
+
+
+def distance_to_convex_hull_boundary(
+    ds: xr.Dataset, ds_reference: xr.Dataset, grid_index_dim: str = "grid_index"
+) -> xr.DataArray:
+    """
+    For all points in `ds` that are external to the convex hull of the points in
+    `ds_reference`, calculate the minimum distance to the convex hull boundary.
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+        The dataset for which to calculate the distance.
+    ds_reference : xarray.Dataset
+        The reference dataset from which to calculate the convex hull boundary.
+    grid_index_dim : str, optional
+        The name of the grid index dimension in `ds` and `ds_reference`.
+
+    Returns
+    -------
+    xarray.DataArray
+        The minimum distance to the convex hull boundary.
+    """
+    # rename the grid index dimension in ds_reference to avoid conflicts (since
+    # the grid index dimension otherwise has the same name in both datasets,
+    # and later we will want to find the minimum distance for each point in ds
+    # to all points in ds_reference)
+    ds_reference_separate_gridindex = ds_reference.rename(
+        {grid_index_dim: "grid_index_ref"}
+    )
+
+    # create a mask from the convex hull of ds_reference for the grid points in ds
+    da_ch_mask = create_convex_hull_mask(
+        ds=ds, ds_reference=ds_reference_separate_gridindex
+    )
+
+    # only consider points that are external to the convex hull
+    ds_exterior = ds.where(~da_ch_mask, drop=True)
+
+    da_xyz = _latlon_to_unit_sphere_xyz(ds_exterior.lat, ds_exterior.lon)
+    da_xyz_ref = _latlon_to_unit_sphere_xyz(
+        ds_reference_separate_gridindex.lat, ds_reference_separate_gridindex.lon
+    )
+
+    def calc_dist(da_pt_xyz):
+        dotproduct = np.dot(da_xyz_ref, da_pt_xyz.T)
+        val = np.min(np.arccos(dotproduct))
+        return val
+
+    da_mindist_to_ref = xr.apply_ufunc(
+        calc_dist,
+        da_xyz,
+        input_core_dims=[["xyz"]],
+        output_core_dims=[[]],
+        vectorize=True,
+    )
```

Comment:
I'm scared about the scaling of this, being $O(NM)$ (number of points in both sets). I understand that this can be parallelized, but we can do better algorithmically.

Once we have computed the convex hull, why don't we compute the distance to this polygon rather than to all points? This does not seem directly implemented in `spherical_geometry`, but maybe https://spherely.readthedocs.io/en/latest/_api_generated/spherely.distance.html#spherely.distance does this correctly?

Alternatively, if we can't measure the distance to the hull, can we do some kd-tree for this, e.g. with https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html + haversine distance?

Reply:
We can get the points defining the polygon from `chull_lam_lon, chull_lam_lat = list(chull_lam.to_lonlat())[0]`. Then we only need to compute minimum distance from (a reasonable subset of) the points in the border to the arcs defined by the polygon segments defining the hull. I think the minimum distance can be computed with something like:
```
def shortest_distance_to_arc(P_lat, P_lon, A_lat, A_lon, B_lat, B_lon, radius=6371):
    P = lat_lon_to_cartesian(P_lat, P_lon)
    A = lat_lon_to_cartesian(A_lat, A_lon)
    B = lat_lon_to_cartesian(B_lat, B_lon)
    
    # Calculate normal vector to the plane of the great circle
    n = np.cross(A, B)
    n = n / np.linalg.norm(n)  # Normalize
    
    # Project P onto the plane
    P_proj = P - np.dot(P, n)[:, np.newaxis] * n
    
    # Normalize to get Q on the sphere's surface
    Q = P_proj / np.linalg.norm(P_proj, axis=1)[:, np.newaxis]
    
    # Calculate the angle between P and Q
    theta = np.arccos(np.clip(np.sum(P * Q, axis=1), -1, 1))
    
    # Check if Q is between A and B
    Q_between = (np.dot(np.cross(A, Q), n) >= 0) & (np.dot(np.cross(Q, B), n) >= 0)
    
    # Calculate distances PA and PB
    PA = np.arccos(np.clip(np.dot(P, A), -1, 1))
    PB = np.arccos(np.clip(np.dot(P, B), -1, 1))
    
    # Choose the appropriate distance
    distances = np.where(Q_between, theta, np.minimum(PA, PB))
    
    return radius * distances
```
Some technical admin is needed to loop over the border point and the segments and find the minimum. Worth the effort?

Reply:
Yes, agreed. It would be good to optimise this. I was going for creating a working implementation with what we have first, and then a later PR could do the extra work to optimise parts of this.

Reply:
I was thinking that I would take quick look to see if I could get the just the coordinates of the polygon though (as you also suggested) and implement that here if it turns out to be easy

Reply:
I think that just getting the coordinates of the points that make up the polygon and restricting the distance computation to those would be an optimization that's completely sufficient. Great if you can give that a try. If it ends up being too complicated I am also fine with just optimizing this later.